src/base/UnorderedNonce.sol

```diff
@@ -0,0 +1,30 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.24;
+
+contract UnorderedNonce {
+    error NonceAlreadyUsed();
+
+    mapping(address owner => mapping(uint256 word => uint256 bitmap)) public nonces;
+
+    /// @notice Consume a nonce, reverting if its already been used
+    /// @param owner address, the owner/signer of the nonce
+    /// @param nonce uint256, the nonce to consume
+    function _useUnorderedNonce(address owner, uint256 nonce) internal {
```

Comment:
would love tests of this function in particular - does it only set the bit in question, does it revert if the nonce was already taken?
if you look in core at `test_fuzz_flipTick`, something like that to check that `_useUnorderedNonce` always sets the right bit, doesnt set others, and reverts always when it was already set?

Reply:
I think there are these exact fuzz tests in permit2 if you want to also transfer them over @saucepoint